lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -122,31 +122,82 @@ static Type getHandshakeBundleDataType(BundleType bundle) {
     return NoneType::get(bundle.getContext());
 }
 
-static Type getHandshakeDataType(Operation *op) {
-  if (auto memOp = dyn_cast<MemoryOp>(op))
-    return memOp.getMemRefType().getElementType();
-
-  else if (auto sinkOp = dyn_cast<SinkOp>(op)) {
-    // As SinkOp only has one argument, which at this stage is already converted
-    // to a bundled FIRRTLType, here we convert it back to a normal data type.
-    // Is there a better way to do this?
-    auto type = sinkOp.getOperand().getType().cast<BundleType>();
-    return getHandshakeBundleDataType(type);
+/// Extracts the type of the data-carrying type of opType. If opType is a
+/// bundle, getHandshakeBundleDataType extracts the data-carrying type, else,
+/// assume that opType itself is the data-carrying type.
+static Type getOperandDataType(Value op) {
+  auto opType = op.getType();
+  if (auto bundleType = opType.dyn_cast<BundleType>(); bundleType)
+    return getHandshakeBundleDataType(bundleType);
+  return opType;
+}
+
+/// Filters types of type TypeToFilter from the input.
+template <typename TypeToFilter>
+static SmallVector<Type> filterTypes(ArrayRef<Type> input) {
```

Comment:
Could this just be `filterNonNoneTypes` or something like that? Seems like this is only parameterized by NoneType.